Make KestrelConfigurationLoader.Load safe to re-call

amcasey opened the pull request:
If it is called again and nothing has changed, it should do nothing.

Part of #45801

amcasey committed:
Make KestrelConfigurationLoader.Load safe to re-call

If it is called again and nothing has changed, it should do nothing.

Part of #45801

halter73 commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -234,25 +235,27 @@ internal void ApplyHttpsDefaults(HttpsConnectionAdapterOptions httpsOptions)
     // Any endpoints that were removed from the last time endpoints were loaded are returned.
     internal (List<ListenOptions>, List<ListenOptions>) Reload()
     {
+        _reloadToken = Configuration.GetReloadToken();
```
```suggestion
        _reloadToken = ReloadOnChange ? Configuration.GetReloadToken() : NullChangeToken.Singleton;
```

We should not grab the token on load if `KestrelConfigurationLoader.ReloadOnChange` is false. That's what the is `DoesNotReloadOnConfigurationChangeByDefault` test is about. Normally it's true if you have a "default" host like `WebApplication`.

You'll need to add using `Microsoft.Extensions.FileProviders` too. Not the best namespace there.

amcasey replied:
I didn't change the existing `IChangeToken` handling, so I believe that test is still passing.  I intentionally decoupled rebind-on-change from determine-if-reloading-would-be-redundant, but I'm open to discussion.

halter73 commented on src/Servers/Kestrel/Core/test/KestrelServerTests.cs:
```diff
@@ -958,7 +958,7 @@ public async Task DoesNotReloadOnConfigurationChangeByDefault()
         mockTransportFactory.Verify(f => f.BindAsync(new IPEndPoint(IPAddress.IPv6Any, 5001), It.IsAny<CancellationToken>()), Times.Once);
 
-        mockConfig.Verify(c => c.GetReloadToken(), Times.Never);
+        mockConfig.Verify(c => c.GetReloadToken(), Times.Once); // Grabbed on Load
```
```suggestion
```

amcasey replied:
Done.

amcasey committed:
Fix copy-paste error

halter73 commented on src/Servers/Kestrel/Kestrel/test/KestrelConfigurationLoaderTests.cs:
```diff
@@ -1224,6 +1226,124 @@ public void Reload_RerunsNamedEndpointConfigurationOnChange()
+        }).Build();
+
+        var mockReloadToken = new Mock<IChangeToken>();
+        mockReloadToken.SetupGet(t => t.HasChanged).Returns(changeBeforeInitialLoad);
```
Why are we parameterizing this return value when below we assert `mockReloadToken.VerifyGet(t => t.HasChanged, Times.Never);` right before reconfiguring it to return `changeAfterInitialLoad`? It couldn't possibly make a difference given we verify it's never called.

amcasey replied:
I agree that it has no effect on the behavior - I thought it was easier to reason about with a more complete implementation.

halter73 replied:
I think it makes things more confusing because it makes you wonder why the change token is even being checked that early. And it ends up adding two redundant test cases.

amcasey replied:
It's true that the tests are redundant with the current product code.  My goal in writing the test was to warn people when they make subtly incorrect refactorings.  If the goal is only to test the code as it exists today, then I can get rid of the check.

amcasey replied:
Removed.

Tratcher commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -234,25 +235,27 @@ internal void ApplyHttpsDefaults(HttpsConnectionAdapterOptions httpsOptions)
     public void Load()
     {
-        if (_loaded)
+        if (_reloadToken is null || _reloadToken.HasChanged)
```
Should this honor `!ReloadOnChange` and not reload?

amcasey replied:
I understood that property to control _implicit_ reloads.  Surely, if you call `Load`, you actually want it to reflect the latest values.

halter73 replied:
I'm a little hesitant to start using the token in any scenario where we didn't at all before because you never know what `GetReloadToken()` implementations might throw or otherwise misbehave. If it's our view that an extra call to `Load()` is an explicit signal, maybe we should always reread configuration in `Load()` without checking the token. 

I think the safest solution is to keep the `_loaded` bool and continue to no-op after the first call to the public `Load()` method.  `UseHttps()` and `KestrelServerImpl` can call internal methods that check the `_reloadToken` only in the `ReloadOnChange` case. Having `UseHttps()` call the `EndpointsToAdd` callbacks and potentially raise exceptions from them would be weird anyway.

halter73 replied:
I'm going to approve this because it's not a huge deal, but I'd prefer we not call `GetReloadToken()` when we do not need to.

I think the public `Load()` method should continue no-op after the first call for simplicity's sake. I haven't heard anyone ask for it not to. And if we do that, there is no reason to capture the token in the `!ReloadOnChange` case. After we fix `UseHttps()` to eagerly load config by calling some other internal method, there will be no real reason to ever call `Load()` anyway.

halter73 commented on src/Servers/Kestrel/Kestrel/test/KestrelConfigurationLoaderTests.cs:
```diff
@@ -1224,6 +1226,124 @@ public void Reload_RerunsNamedEndpointConfigurationOnChange()
+        serverOptions.ConfigurationLoader.Load();
+
+        mockReloadToken.VerifyGet(t => t.HasChanged, Times.AtLeastOnce);
+        mockConfig.Verify(c => c.GetSection(It.IsAny<string>()), changeAfterInitialLoad ? Times.AtLeastOnce : Times.Never);
```
Can we also modify the endpoints between loads assert that `serverOptions.ConfigurationBackedListenOptions` is what we expect after each load?

amcasey replied:
I'm having trouble imagining a failure mode where we get the right number of endpoints but the wrong values, but I'm willing to add additional checks if you think it's important.

halter73 replied:
A check that we have the expected number of endpoints is fine. I don't see that in this test though. The number of endpoints also doesn't change after the first load. The easiest way verify we get the right number is probably to assert the `serverOptions.ConfigurationBackedListenOptions.Count`.

The `MultipleLoads_AddEndpoints` test below verifies that the callbacks are called the correct number of times, but even that doesn't necessarily mean the right number of endpoints have been added. It's easy to verify both.

amcasey replied:
I had assumed `KestrelConfigurationLoader.LocalhostEndpoint` had existing test coverage, but I've updated this test.  Note that these endpoints actually go in `CodeBasedListenOptions`, rather than `ConfigurationBackedListenOptions`.

amcasey committed:
Update tests based on PR feedback

halter73 commented on src/Servers/Kestrel/Core/src/KestrelConfigurationLoader.cs:
```diff
@@ -234,25 +235,27 @@ internal void ApplyHttpsDefaults(HttpsConnectionAdapterOptions httpsOptions)
 
         foreach (var action in EndpointsToAdd)
         {
             action();
```
Can we add a test for what happens a user-defined `Action<ListenOptions>` throws out of here?

Previously, if someone manually called `Load()` and swallowed the exception, we'd bind to everything defined before the throwing-action. I don't think it's important to preserve that behavior. Failing to bind is fine, but it'd be good to have a test for it.

It could lead to some really weird conflicts if there are any actions that throw non-deterministically, so it's probably best to avoid ever rerunning an action.

amcasey replied:
I've added such a test and added a flag to prevent re-runs.

amcasey committed:
Factor LoadInternal and ProcessEndpointsToAdd out of Load

...so that `Load` can retain its current behavior.  Also, only consume the change token when `ReloadOnChange` is true.

amcasey committed:
Remove unused using

halter73 commented on src/Servers/Kestrel/Core/test/KestrelServerTests.cs:
```diff
@@ -958,8 +958,6 @@ public async Task DoesNotReloadOnConfigurationChangeByDefault()
         mockTransportFactory.Verify(f => f.BindAsync(new IPEndPoint(IPAddress.IPv6Any, 5000), It.IsAny<CancellationToken>()), Times.Once);
         mockTransportFactory.Verify(f => f.BindAsync(new IPEndPoint(IPAddress.IPv6Any, 5001), It.IsAny<CancellationToken>()), Times.Once);
 
-        mockConfig.Verify(c => c.GetReloadToken(), Times.Never);
```
Can we add this assertion back now?

amcasey replied:
Done.

halter73 reviewed (commented):
Looks great!

amcasey committed:
Revert KestrelServerTests
